docs/proposals/1001-inference-scheduling-objective/README.md

```diff
@@ -0,0 +1,257 @@
+
+- Drop the `InferenceModel` API, but no changes to the InferencePool API
+- Replace the API with `InferenceSchedulingObjective`. The API will define `scheduling objectives` for matching requests. The inference-scheduler (run by the EPP) will be the primary actuator of this API.
+- We will keep the Criticality field, with the intent to add more serving objectives (e.g., latency SLOs)
+- Request matching beyond model name to include headers. This allows defining different scheduling policies for different request flows (apps or users) while targeting the same model. The semantics should be adopted from HTTPRouteMatch as defined in HTTPRouteRules.
+- A default InferenceSchedulingObjective per InferencePool will be included as a fallback policy when no one matches the request. These defaults can be adjusted.
+- Traffic splitting is not part of the InferenceSchedulingObjective API. Traffic splitting is not an endpoint scheduling objective, it is a request routing objective. As we describe below, with some creativity, we can offload traffic splitting to  HTTPRoute.
+  - An intended side effect of this is that users will more easily be able to define different scheduling policies for the same target models, something that required some shenanigans with the current API (2 inferenceModels with distinct `modelNames` both pointing to the same target model).
+
+```golang
+type InferenceSchedulingObjectivesSpec struct {
+  // Match defines what requests this objective applies to.
+  Match
+
+  // Criticality defines how important it is to serve the requests that match this objective
+  // compared to requests that match other objectives.
+  // Criticality impacts how traffic is handled in resource constrained situations. It handles this 
+  // by queuing or rejecting requests of lower criticality. Objectives of an equivalent Criticality 
+  // will fairly share resources over throughput of tokens. In the future, the metric used to 
+  // calculate fairness, and the proportionality of fairness will be configurable.
+  //
+  //
+  // Default values for this field will not be set, to allow for future additions of a new field that 
+  // may 'one of' with this field.
+  // Any implementations that may consume this field may treat an unset value as the 
+  // 'Standard' range.
+  Criticality *Criticality 
+
+  // Future scheduling objectives, like SLOs.
+
+  // PoolRef is a reference to the inference pool, the pool must exist in the same namespace.
+  PoolRef PoolObjectReference
+}
+type Match struct {
+  // Only one of the following can be set.
+
+  // HTTPMatches is a list of http requests matchers, the list of matchers are ORed.
+  HTTPMatches []HTTPMatch
+  // GRPCMatches is a list of gRPC requests matchers, the list of matchers are ORed.
+  GRPCMatches []GRPCMatch
+}
+
+// HTTPMatch is an http matching rule. The rules are ANDed.
+type HTTPMatch struct {
+  // ModelName matches against the model name in the body as per OpenAI protocol
+  ModelName *string
```

Comment:
if one wants to specify multiple models all with app: prod, he needs to define multiple HttpMatch entries, and repeat the same headers in all. 
does it make sense to make this field an array of model names?  (with a comment that list of models is Or’ed)

Reply:
ok. I'll wait for the change to comment. 
using an array also keeps the modelName optional (array can be empty) :).